src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,154 @@
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        MyLinkedList.Node<T> newValue = getNodeByIndex(index);
+        T oldValue = newValue.item;
+        newValue.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T oldValue = currentNode.item;
+        unlink(currentNode);
+        return oldValue;
     }
 
     @Override
-    public boolean remove(T object) {
+    public boolean remove(T element) {
+        for (MyLinkedList.Node<T> removedNode = first; removedNode
+                != null; removedNode = removedNode.next) {
+            if (removedNode.item == element
+                    || (element != null && element.equals(removedNode.item))) {
+                unlink(removedNode);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private MyLinkedList.Node<T> getNodeByIndex(int index) {
+        Node<T> foundedNode;
+        if (index < (size / 2)) {
+            foundedNode = first;
+            for (int i = 0; i < index; i++) {
+                foundedNode = foundedNode.next;
+            }
+        } else {
+            foundedNode = last;
+            for (int i = size - 1; i > index; i--) {
+                foundedNode = foundedNode.prev;
+            }
+        }
+        return foundedNode;
+    }
+
+    private void unlink(Node<T> removedNode) {
+        MyLinkedList.Node<T> next = removedNode.next;
+        MyLinkedList.Node<T> prev = removedNode.prev;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            removedNode.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            removedNode.next = null;
+        }
+        removedNode.item = null;
+        size--;
+    }
+
+    private void linkBefore(T element, MyLinkedList.Node<T> afterNode) {
+        final MyLinkedList.Node<T> prev = afterNode.prev;
+        final MyLinkedList.Node<T> newNode = new MyLinkedList.Node<>(prev, element, afterNode);
+        if (prev == null) {
+            first = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        afterNode.prev = newNode;
+        size++;
+    }
+
+    private void checkElementIndex(int index) {
```

Comment:
`checkElementIndex` and `checkPositionIndex` are absolutely the same. Use only one of these methods. Also, rename it as `checkIndex` for better understanding.

Reply:
they is`not the same because first checkElementIndex need to have index < size, second index <= size